src/dali/abstract_pair_converter_plugin.py

```diff
@@ -88,11 +109,61 @@ def get_or_set_vertex(self, name: str) -> Vertex[ValueType]:
         self.add_vertex(new_vertex)
         return new_vertex
 
-    def add_neighbor(self, vertex_name: str, neighbor_name: str, weight: float = 0.0) -> None:
+    def is_optimized(self, asset: str) -> bool:
+        return asset in self.__optimized_assets
+
+    @property
+    def optimized_assets(self) -> Set[str]:
+        return self.__optimized_assets
+
+    # Optimization contains a dict with a key of the optimized asset and a value of a dict with the optimized weights for each neighbor
+    # Optimized assets are tracked to prevent requesting prices for unoptimized assets
+    # Negative weights will get deleted.
+    def clone_with_optimization(self, optimization: Dict[str, Dict[str, float]]) -> "MappedGraph[ValueType]":
+        for optimized_asset in optimization.keys():
+            if optimized_asset in self.__optimized_assets:
+                raise RP2ValueError(
+                    f"INTERNAL ERROR: You are trying to optimize an asset that has already been optimized. "
+                    f"Use is_optimized() to check if an asset has already been optimized before cloning."
+                    f"Found {optimized_asset} in {self.__optimized_assets}."
+                )
+
+        # wrapping optimized_assets in set() makes a deep copy
+        cloned_mapped_graph: MappedGraph[ValueType] = MappedGraph(optimized_assets=set(self.optimized_assets))
+
+        for original_vertex in self.vertexes:
+            # Add existing neighbors
+            for neighbor in original_vertex.neighbors:
+                neighbor_weight: float
+                optimized: bool = False
+                if original_vertex.name in set(optimization.keys()):
```

Comment:
Assign the set to a variable outside of the loop, to avoid rebuilding it at every iteration.